File: face/geomajas-face-dojo/client/src/main/resources/geomajas/map/search/SearchCriterion.js
/*
 * This is part of Geomajas, a GIS framework, http://www.geomajas.org/.
 *
 * Copyright 2008-2011 Geosparc nv, http://www.geosparc.com/, Belgium.
 *
 * The program is available in open source according to the GNU Affero
 * General Public License. All contributions in this program are covered
 * by the Geomajas Contributors License Agreement. For full licensing
 * details, see LICENSE.txt in the project root.
 */

dojo.provide("geomajas.map.search.SearchCriterion");
dojo.declare("SearchCriterion", null, {

	/**
	 * @fileoverview One criterion for a search request.
	 * @class One criterion for a search request.
	 * @author Pieter De Graef
	 *
	 * @constructor
	 * @param attributeDefinition The {@link AttributeDefinition} object we
	 *                            are to search for.
	 * @param operator The search operator as a string (like, =, <, ...).
	 * @param value The value literal. Can be complex in case the attribute is
	 *              an association.
	 */
	constructor : function (attributeDefinition, operator, value) {
		this.attributeDefinition = attributeDefinition;
		this.operator = operator;
		this.value = value;
	},

	//-------------------------------------------------------------------------
	// SearchCriterion specific functions:
	//-------------------------------------------------------------------------

	/**
	 * Transforms the criterion to a string.
	 */
	toString : function () {
		return this.attributeDefinition.getName()+" "+this.operator+" "+this.value;
	},

	/**
	 * Prepares this searchcriterion for sending to the server through JSON.
	 * The serverside representation of this class, accept attributeName,
	 * operator and value, all as string. This means that in case of an
	 * association, we need to send the primitive value!
	 * @return Returns a JSON serializable object.
	 */
	toJSON : function () {
		var primitiveValue = this.value;
		var attributeName = this.attributeDefinition.getName();
		if (this.attributeDefinition.isAssociation()) {
			attributeName += "." + this.attributeDefinition.getAssociationIdentifierName();
			primitiveValue = this.value[this.attributeDefinition.getAssociationIdentifierName()];
		}
		return {
			javaClass : "org.geomajas.layer.feature.SearchCriterion",
			attributeName : attributeName,
			operator : this.operator,
			value : primitiveValue
		};
	},



	//-------------------------------------------------------------------------
	// Getters and setters:
	//-------------------------------------------------------------------------

	getAttributeDefinition : function () {
		return this.attributeDefinition;
	},

	setAttributeDefinition : function (attributeDefinition) {
		this.attributeDefinition = attributeDefinition;
	},

	getOperator : function () {
		return this.operator;
	},

	setOperator : function (operator) {
		this.operator = operator;
	},

	getValue : function () {
		return this.value;
	},

	setValue : function (value) {
		this.value = value;
	}
});

File: api/src/main/java/org/geomajas/layer/feature/SearchCriterion.java
/*
 * This is part of Geomajas, a GIS framework, http://www.geomajas.org/.
 *
 * Copyright 2008-2013 Geosparc nv, http://www.geosparc.com/, Belgium.
 *
 * The program is available in open source according to the GNU Affero
 * General Public License. All contributions in this program are covered
 * by the Geomajas Contributors License Agreement. For full licensing
 * details, see LICENSE.txt in the project root.
 */
package org.geomajas.layer.feature;

import java.io.Serializable;

import org.geomajas.annotation.Api;

/**
 * A single criterion DTO that can be used to set up filters and search for features by their alpha numerical
 * attributes.
 * 
 * @author Pieter De Graef
 * @since 1.6.0
 */
@Api(allMethods = true)
public class SearchCriterion implements Serializable {

	private static final long serialVersionUID = 151L;

	private String attributeName;

	private String operator;

	private String value;

	// -------------------------------------------------------------------------
	// Constructors:
	// -------------------------------------------------------------------------

	/**
	 * Default constructor - does nothing.
	 */
	public SearchCriterion() {
	}

	/**
	 * Constructor that initializes all fields with their values.
	 * 
	 * @param attributeName
	 *            Set the name of the alpha numerical attribute on which this criterion operates. This attribute name is
	 *            one of the attributes from the vector layer configuration.
	 * @param operator
	 *            The operator used in the criterion such as '=', '>', ...
	 * @param value
	 *            The value for this criterion as a formatted string.
	 */
	public SearchCriterion(String attributeName, String operator, String value) {
		this.attributeName = attributeName;
		this.operator = operator;
		this.value = value;
	}

	// -------------------------------------------------------------------------
	// Public methods:
	// -------------------------------------------------------------------------

	/**
	 * Convert to human readable form.
	 * 
	 * @return human readable form
	 */
	public String toString() {
		return "(" + attributeName + " " + operator + " " + value + ")";
	}

	// -------------------------------------------------------------------------
	// Getters and setters:
	// -------------------------------------------------------------------------

	/**
	 * Return the operator used in the criterion such as '=', '>', ...
	 * 
	 * @return operator
	 */
	public String getOperator() {
		return operator;
	}

	/**
	 * Set the operator used in the criterion such as '=', '>', ...
	 * 
	 * @param operator
	 *            The criterion operator.
	 */
	public void setOperator(String operator) {
		this.operator = operator;
	}

	/**
	 * Get the name of the alpha numerical attribute on which this criterion operates. This attribute name is one of the
	 * attributes from the vector layer configuration.
	 * 
	 * @return attribute name
	 */
	public String getAttributeName() {
		return attributeName;
	}

	/**
	 * Set the name of the alpha numerical attribute on which this criterion operates. This attribute name is one of the
	 * attributes from the vector layer configuration.
	 * 
	 * @param attributeName
	 *            The name of the alpha numerical attribute.
	 */
	public void setAttributeName(String attributeName) {
		this.attributeName = attributeName;
	}

	/**
	 * Get the value for this criterion as a formatted string.
	 * 
	 * @return value for comparison
	 */
	public String getValue() {
		return value;
	}

	/**
	 * Set the value for this criterion as a formatted string.
	 * 
	 * @param value
	 *            The value for the attribute.
	 */
	public void setValue(String value) {
		this.value = value;
	}
}